File: backend/trace_generator.py
"""
trace_generator.py - Generates synthetic memory access traces.
Patterns: sequential, random, temporal_locality, spatial_locality, thrashing.
"""
from __future__ import annotations
import random
from typing import List


def generate_trace(pattern: str, count: int = 100, address_space: int = 0xFFFF) -> List[int]:
    """
    Generate a list of memory addresses for the given pattern.

    Parameters
    ----------
    pattern      : str  One of sequential | random | temporal | spatial | thrashing
    count        : int  Number of accesses to generate
    address_space: int  Maximum address value
    """
    pattern = pattern.lower()
    rng = random.Random(42)  # fixed seed for reproducibility

    if pattern == "sequential":
        return [(i * 64) % address_space for i in range(count)]

    elif pattern == "random":
        return [rng.randint(0, address_space) & ~0x3F for _ in range(count)]

    elif pattern in ("temporal", "temporal_locality"):
        # 80% of accesses hit a small working set of ~5 addresses
        hot_set = [rng.randint(0, address_space // 4) & ~0x3F for _ in range(5)]
        addresses = []
        for _ in range(count):
            if rng.random() < 0.8:
                addresses.append(rng.choice(hot_set))
            else:
                addresses.append(rng.randint(0, address_space) & ~0x3F)
        return addresses

    elif pattern in ("spatial", "spatial_locality"):
        # Walk through memory in strides of 1 block, occasionally jump
        base = rng.randint(0, address_space // 2) & ~0x3F
        addresses = []
        for _ in range(count):
            addresses.append(base % address_space)
            if rng.random() < 0.1:
                base = rng.randint(0, address_space // 2) & ~0x3F
            else:
                base += 64
        return addresses

    elif pattern == "thrashing":
        # Access a working set just slightly larger than the cache
        # so every access evicts something it will need shortly after
        working_set_size = 16  # should exceed typical cache capacity in blocks
        hot_set = [(i * 64) % address_space for i in range(working_set_size)]
        return [hot_set[i % working_set_size] for i in range(count)]

    else:
        raise ValueError(
            f"Unknown pattern '{pattern}'. "
            "Choose: sequential | random | temporal | spatial | thrashing"
        )

File: frontend/src/lib/simulator/traceGenerator.ts
// src/lib/simulator/traceGenerator.ts
import type { TracePattern } from './types';

// Simple mulberry32 PRNG for deterministic trace generation
function mulberry32(a: number) {
  return function() {
    let t = a += 0x6D2B79F5;
    t = Math.imul(t ^ t >>> 15, t | 1);
    t ^= t + Math.imul(t ^ t >>> 7, t | 61);
    return ((t ^ t >>> 14) >>> 0) / 4294967296;
  }
}

export function generateTrace(pattern: TracePattern, count: number, addressSpace = 0xFFFF, seed = 42): number[] {
  const rng = mulberry32(seed);
  const randInt = (min: number, max: number) => Math.floor(rng() * (max - min + 1)) + min;
  
  const addresses: number[] = [];
  const blockSize = 64; // assume 64B for generation alignment

  if (pattern === 'sequential') {
    for (let i = 0; i < count; i++) {
      addresses.push((i * blockSize) % addressSpace);
    }
  } 
  
  else if (pattern === 'random') {
    for (let i = 0; i < count; i++) {
      const addr = randInt(0, addressSpace);
      addresses.push(addr - (addr % blockSize)); // align
    }
  }
  
  else if (pattern === 'temporal') {
    // 80% hits a small hot set of 5 blocks
    const hotSet = Array.from({ length: 5 }, () => {
      const a = randInt(0, Math.floor(addressSpace / 4));
      return a - (a % blockSize);
    });
    for (let i = 0; i < count; i++) {
      if (rng() < 0.8) {
        addresses.push(hotSet[randInt(0, hotSet.length - 1)]);
      } else {
        const addr = randInt(0, addressSpace);
        addresses.push(addr - (addr % blockSize));
      }
    }
  }
  
  else if (pattern === 'spatial') {
    // Walk in strides, occasionally jump
    let base = randInt(0, Math.floor(addressSpace / 2));
    base = base - (base % blockSize);
    for (let i = 0; i < count; i++) {
      addresses.push(base % addressSpace);
      if (rng() < 0.1) {
        base = randInt(0, Math.floor(addressSpace / 2));
        base = base - (base % blockSize);
      } else {
        base += blockSize;
      }
    }
  }
  
  else if (pattern === 'thrashing') {
    // Working set just slightly larger than typical cache
    const workingSetSize = 16; 
    const hotSet = Array.from({ length: workingSetSize }, (_, i) => (i * blockSize) % addressSpace);
    for (let i = 0; i < count; i++) {
      addresses.push(hotSet[i % workingSetSize]);
    }
  }
  
  else if (pattern === 'mixed') {
    // A mix of sequential, temporal, and random
    for (let i = 0; i < count; i++) {
      const r = rng();
      if (r < 0.3) addresses.push((i * blockSize) % addressSpace);
      else if (r < 0.8) addresses.push(randInt(0, 5) * blockSize); // tight temporal
      else {
        const a = randInt(0, addressSpace);
        addresses.push(a - (a % blockSize));
      }
    }
  }

  return addresses;
}
